server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface QueryCancellable {
+
+        void checkCancelled();
+    }
+
+    /**
+     * Wraps an {@link IndexReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableDirectoryReader extends FilterDirectoryReader {
+
+        private final Holder<QueryCancellable> cancellableHolder;
+
+        private CancellableDirectoryReader(DirectoryReader in, Holder<QueryCancellable> cancellableHolder) throws IOException {
```

Comment:
this constructor seems to always be called with a `Holder` around a `null`, so let's remove it from the ctor arguments and just set `cancellableHolder = new Holder<>(null);`?

Reply:
The issue is that we need to pass it in the ctor because we need to use it in the constructor: https://github.com/elastic/elasticsearch/pull/52822/files/9695114f3028f98ae59dc5ff7481aab3251bce9f#diff-913ad694a7d744ee93ae1dac48d67b0eR368
so that it's passed down the object tree properly.

Reply:
Ah, I see. Nevermind then.

Reply:
I also wonder if we can replace the `Holder` with a mutable `QueryTimeout` implementation:
Something like:
`````
   private static class MutableQueryTimeout implements QueryTimeout {
        private final Set<Runnable> runnables = new HashSet<>();

        Object add(Runnable action) {
            if (runnables.add(action) == false) {
                throw new IllegalArgumentException("already added");
            }
            return action;
        }

        boolean remove(Runnable action) {
            return runnables.remove(action);
        }

        @Override
        public boolean shouldExit() {
            for (Runnable timeout : runnables) {
                timeout.run();
            }
            return false;
        }

        @Override
        public boolean isTimeoutEnabled() {
            return runnables.size() > 0;
        }
    }
`````
This way you can expose a simple `QueryTimeout` in the wrapped reader and keep the logic simple there ? 
The `ContextIndexSearcher` could then expose the `addQueryTimeout`, `removeQueryTimeout` directly for the `QueryPhase` modification.